File: epi_judge_java/epi/IsListPalindromic.java
package epi;

import epi.test_framework.EpiTest;
import epi.test_framework.GenericTest;

public class IsListPalindromic {

    @EpiTest(testDataFile = "is_list_palindromic.tsv")
    public static boolean isLinkedListAPalindrome(ListNode<Integer> head) {
        if (head == null || head.next == null) return true;

        var fast = head.next;
        var slow = head;

        while (fast != null && fast.next != null) {
            fast = fast.next.next;
            slow = slow.next;
        }

        fast = slow.next;
        slow.next = null;
        fast = reverse(fast);

        while (head != null && fast != null) {
            if (!head.data.equals(fast.data)) return false;
            head = head.next;
            fast = fast.next;
        }

        return true;
    }

    private static ListNode<Integer> reverse(ListNode<Integer> head) {
        var runner = head;

        while (runner != null) {
            var next = runner.next;
            runner.next = head;
            head = runner;
            runner = next;
        }

        return head;
    }

    public static void main(String[] args) {
        System.exit(
                GenericTest
                        .runFromAnnotations(args, "IsListPalindromic.java",
                                new Object() {
                                }.getClass().getEnclosingClass())
                        .ordinal());
    }
}

File: epi_judge_cpp/is_list_palindromic.cc
#include "list_node.h"
#include "test_framework/generic_test.h"

int count = 0;

bool IsLinkedListAPalindrome(shared_ptr<ListNode<int>> list) {
    count++;
    if (list == nullptr || list->next == nullptr) return true;

    int length = 0;
    auto runner = list;

    while (runner != nullptr) {
        length++;
        runner = runner->next;
    }

    shared_ptr<ListNode<int>> reversedDummy(new ListNode<int>);
    int reverseStartIndex = (length + 1) / 2;
    int half = length / 2;

    runner = list;
    for (int i = 0; i < reverseStartIndex; ++i) {
        runner = runner->next;
    }

    reversedDummy->next = runner;
    while (runner->next != nullptr) {
        auto temp = runner->next;
        runner->next = runner->next->next;
        temp->next = reversedDummy->next;
        reversedDummy->next = temp;
    }

    runner = list;
    auto reversedRunner = reversedDummy->next;

    for (int i = 0; i < half; ++i) {
        if (runner->data != reversedRunner->data) return false;
        runner = runner->next;
        reversedRunner = reversedRunner->next;
    }

    return true;
}

int main(int argc, char *argv[]) {
    std::vector<std::string> args{argv + 1, argv + argc};
    std::vector<std::string> param_names{"L"};
    return GenericTestMain(args, "is_list_palindromic.cc",
                           "is_list_palindromic.tsv", &IsLinkedListAPalindrome,
                           DefaultComparator{}, param_names);
}
